### ISSUE TITLE
IQ metadata

### ISSUE THREAD

**Issue by srcejon** [f4exb/sdrangel#1143]
While updating the APT & ADS-B demods for the 3D Map, it became apparent that it would be useful to have a timestamp for when the first sample in a block of IQ data was sampled, so the demod can then work out the time of any sample in that block.

For APT, this is useful in aligning each row of pixels with the location of the satellite, so that it can be projected on to the correct coordinates on to the map.

For ADS-B, accurate timing can help give smoother motion of the aircraft on the 3D map. The previous version of the ADS-B demod was creating the receive timestamp for each frame based on when the frame was demodulated, but this resulted in jerky motion as the timestamps were quite "bursty" (presumably due to a combination of the buffering and OS scheduling). I've improved this a bit by creating the timestamp when the demod sink first receives the sample, but I think it would be better if this timestamp was created further up the chain.

It seems there are two delays to consider - the delay from the SDR to SDRangel - and then the delay from when SDRangel first receives the sample to the demod.

For the delay from the SDR to SDRangel, some SDRs have meta data that can help:

- The USRP devices include a time_spec for the first sample in uhd::rx_metadata_t structure. 
- Lime devices have GetTimestamp in lms_stream_meta_t

Some other SDRs don't have this though. However, for these, we can probably still help with the delay through SDRangel, by creating a timestamp manually. One possible way to do this, would be to record the current system time when the first sample is received after acquisition is started - then increment this timestamp based on the number of samples received * the sample rate.

**Comment by srcejon** [comment]
A signal giving the time of the first sample after acquisition starts might be good enough. It's perhaps a reasonable constraint to say timing will only have it's highest accuracy if the user doesn't change the sample rate and there are no buffer overflows etc. I will experiment and try to and get a better idea of what the delays are and how beneficial it might be.

**Comment by f4exb** [comment]
Above commit implements timing signals in the `SampleSinkFifo`. This FIFO is mainly used in the I/Q samples acquisition after decimation. 

The `written` signal has two parameters: `void written(int nsamples, qint64 timestamp);`
  - `nsamples` is the number of samples written to the FIFO since the last signal was emitted
  - `timestamp` is the elapsed time in nanoseconds since the main elapsed time timer in `MainCore` has been started (more on that later)

The rate at which the signal is emitted depends on the rate the device sends blocs of data. It does not depend on decimation. As decimation increases the number of samples written to the FIFO decreases by the same amount. In order to throttle the signal emissions a rate divider can be provided by the `setWrittenSignalRateDivider` method. Thus the signal will be emitted only once every this many writes while the number of samples is accumulated and the total provided in the `nsamples` parameter of the signal. For example for RTLSDR the divider is set to 32 to match approximately the rate of a HackRF of ~128ms. Other devices have not been experimented with yet. 

As mentioned earlier `MainCore` will start an elapsed timer in its constructor thus it will run ans long as the application runs. Its value is returned by `MainCore::instance()->getElapsedNsecs()`. This is the number of nanoseconds since the timer was started and while it has nanosecond precision the accuracy of the valyue depends on the system. Roughly at the same time this timer is started the millisecond epoch is recorded and can be returned by `MainCore::instance()->getStartMsecsSinceEpoch()`. This can help relate the events to the absolute time.

I have not included an example of usage in the commit. However if you would like to implement a listener (slot) in the AM demodulator you would proceed like this:

Create a `qint64 m_lastTs;` member to record the last timestamp to get the elapsed time between `written` signal ticks that you intialize to zero in the constructor. You then implement a `void handleWrittenToFifo(int nsamples, qint64 timestamp);` in the private slots and connect it to the `written` signal in the AM demod constructor like this: 

```c++
    QObject::connect(
        m_deviceAPI->getSampleSource()->getSampleFifo(),
        &SampleSinkFifo::written,
        this,
        &AMDemod::handleWrittenToFifo
    );
```
You can optionally disconnect the signal in the destructor.

The signal handler (slot) body can be something like this. Here it just displays data:

``` c++
void AMDemod::handleWrittenToFifo(int nsamples, qint64 timestamp)
{
    QDateTime dt;
    dt.setMSecsSinceEpoch(MainCore::instance()->getStartMsecsSinceEpoch() + timestamp/1000000);
    qDebug("AMDemod::handleWrittenToFifo: %s: nsamples: %d, dts: %lld sr: %f",
        qPrintable(dt.toString("yyyy.MM.dd hh:mm:ss.zzz")),
        nsamples,
        timestamp - m_lastTs, nsamples*1e9 / (timestamp - m_lastTs)
    );
    m_lastTs = timestamp;
}
```
An extract of the logs prints like this:
```
2022-03-15 22:14:36.412 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:36.411: nsamples: 16384, dts: 109166604 sr: 150082.528902
2022-03-15 22:14:36.522 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:36.521: nsamples: 16384, dts: 109375251 sr: 149796.227668
2022-03-15 22:14:36.631 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:36.630: nsamples: 16384, dts: 109367802 sr: 149806.430233
2022-03-15 22:14:36.740 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:36.739: nsamples: 16384, dts: 109197004 sr: 150040.746539
2022-03-15 22:14:36.849 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:36.849: nsamples: 16384, dts: 109111325 sr: 150158.565117
2022-03-15 22:14:36.958 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:36.958: nsamples: 16384, dts: 109201601 sr: 150034.430356
2022-03-15 22:14:37.068 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.067: nsamples: 16384, dts: 109186814 sr: 150054.749285
2022-03-15 22:14:37.177 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.176: nsamples: 16384, dts: 109394219 sr: 149770.254313
2022-03-15 22:14:37.286 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.285: nsamples: 16384, dts: 108963376 sr: 150362.448388
2022-03-15 22:14:37.396 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.395: nsamples: 16384, dts: 109570377 sr: 149529.466345
2022-03-15 22:14:37.505 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.504: nsamples: 16384, dts: 109025279 sr: 150277.074732
2022-03-15 22:14:37.614 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.613: nsamples: 16384, dts: 109226461 sr: 150000.282441
2022-03-15 22:14:37.723 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.722: nsamples: 16384, dts: 109171834 sr: 150075.339029
2022-03-15 22:14:37.832 (D) AMDemod::handleWrittenToFifo: 2022.03.15 22:14:37.832: nsamples: 16384, dts: 109345147 sr: 149837.468324
```
One can notice that the timestamp of the log and the calculated timestamp from the FIFO events lie within the same millisecond. The original sample rate taking decimation into account is 150 kS/s. The calculated sample rate shown in "sr" fluctuates about this value.

These are the original settings:
![image](https://user-images.githubusercontent.com/6192319/158477638-61bba8fd-e0c2-4eca-9e5e-705a74dd421f.png)

**Comment by f4exb** [comment]
Implemented in v6.20.2
